pkg/autoscaler/autoscaler/autoscaler_manager.go

```diff
@@ -100,8 +101,59 @@ func (am *autoScalerManager) Sync(tac *v1alpha1.TidbClusterAutoScaler) error {
 	return am.updateAutoScaling(oldTc, tac)
 }
 
+func (am *autoScalerManager) syncExternal(tc *v1alpha1.TidbCluster, tac *v1alpha1.TidbClusterAutoScaler, component v1alpha1.MemberType) error {
+	switch component {
+	case v1alpha1.TiDBMemberType:
+		targetReplicas, err := query.ExternalService(tc, v1alpha1.TiDBMemberType, tac.Spec.TiDB.External.Endpoint, am.kubecli)
+		if err != nil {
+			klog.Errorf("tac[%s/%s] 's query to the external endpoint got error: %v", tac.Namespace, tac.Name, err)
+			return err
+		}
+
+		if tc.Spec.TiDB.Replicas == targetReplicas {
+			return nil
+		}
+
+		updated := tc.DeepCopy()
+		updated.Spec.TiDB.Replicas = targetReplicas
+		if _, err = am.tcControl.UpdateTidbCluster(updated, &updated.Status, &tc.Status); err != nil {
+			return err
+		}
+	case v1alpha1.TiKVMemberType:
+		targetReplicas, err := query.ExternalService(tc, v1alpha1.TiKVMemberType, tac.Spec.TiKV.External.Endpoint, am.kubecli)
+		if err != nil {
+			klog.Errorf("tac[%s/%s] 's query to the external endpoint got error: %v", tac.Namespace, tac.Name, err)
+			return err
+		}
+
+		if tc.Spec.TiKV.Replicas == targetReplicas {
+			return nil
+		}
+
+		updated := tc.DeepCopy()
+		updated.Spec.TiKV.Replicas = targetReplicas
+		if _, err = am.tcControl.UpdateTidbCluster(updated, &updated.Status, &tc.Status); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 func (am *autoScalerManager) syncAutoScaling(tc *v1alpha1.TidbCluster, tac *v1alpha1.TidbClusterAutoScaler) error {
-	// TODO: sync from external endpoints if specified
+	// Sync from external endpoints if specified
+	if tac.Spec.TiDB != nil && tac.Spec.TiDB.External != nil {
+		if err := am.syncExternal(tc, tac, v1alpha1.TiDBMemberType); err != nil {
+			return err
+		}
+	}
+
+	if tac.Spec.TiKV != nil && tac.Spec.TiKV.External != nil {
+		if err := am.syncExternal(tc, tac, v1alpha1.TiKVMemberType); err != nil {
+			return err
+		}
+	}
+
```

Comment:
If users configures `external` for TiDB/TiKV, should we still go through to the following code to call PD API?

Reply:
The https://github.com/pingcap/tidb-operator/blob/master/pkg/autoscaler/autoscaler/util.go#L160 and https://github.com/pingcap/tidb-operator/blob/master/pkg/autoscaler/autoscaler/util.go#L259 functions do not consider the external configuration.

Reply:
If user configures `external` for one component, it will be ignored when constructing PD API requests. If the constructed PD strategy has no rules, then we will skip making PD API requests.

Reply:
The defaulting function and the validating function have been updated

Reply:
> If user configures external for one component, it will be ignored when constructing PD API requests. If the constructed PD strategy has no rules, then we will skip making PD API requests.

Where is the PD API skipped?
What's the behavior of the following cases in PD (Y means exist, N means non-exist):
TiDB rules TiKV rules
Y Y
N Y
Y N
N N